### ISSUE TITLE
Response to simple OPTIONS queries is a 405 Method Not Allowed

### ISSUE THREAD

**Issue by RAnders00** [tokio-rs/axum#963]
### Version

```none
├── axum v0.5.3
│   ├── axum-core v0.2.2
```

### Platform

Windows 11 21H2 (Build 22000.613) 64-bit

### Description

Currently, if you make a simple app with CORS middleware like this:

```rust
let cors = CorsLayer::new()
    .allow_methods(vec![Method::GET, Method::POST])
    .allow_origin(cors::Any);

let app = Router::new()
    .route("/", get(|| async { "Hello, World!" }))
    .layer(cors);
```

...and send that service an empty `OPTIONS` request (`OPTIONS /`), it'll respond with `405 Method Not Allowed`:

```none
> OPTIONS / HTTP/1.1
> Host: localhost:2790

< HTTP/1.1 405 Method Not Allowed
< content-length: 0
< date: Mon, 25 Apr 2022 09:55:06 GMT
```

I'm not sure how strongly whether the relevant HTTP standards/RFCs allow this, but I've just read the MDN article on `OPTIONS` HTTP verb and it seems like a blank `OPTIONS` request should be possible: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS#identifying_allowed_request_methods

> The response then contains an [Allow](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Allow) header that holds the allowed methods:
> 
> ```
> HTTP/1.1 204 No Content
> Allow: OPTIONS, GET, HEAD, POST
> Cache-Control: max-age=604800
> Date: Thu, 13 Oct 2016 11:45:00 GMT
> Server: EOS (lax004/2813)
> ```

So, in short, I'd expect my small axum service with cors middleware to respond to the simple `OPTIONS` request with a response similar to this:

```
HTTP/1.1 204 No Content
allow: GET, POST
date: Mon, 25 Apr 2022 09:55:06 GMT
```

Instead of the `405 Method Not Allowed` shown above.

**Comment by RAnders00** [comment]
I should further add:

Adding an `Origin` header to the request makes the examplary service from above respond with:

```
> OPTIONS / HTTP/1.1
> Host: localhost:2790
> Origin: http://example.com

< HTTP/1.1 200 OK
< allow: GET,HEAD
< content-length: 0
< date: Mon, 25 Apr 2022 10:10:27 GMT
```

**Comment by RAnders00** [comment]
I've had more of a dig into `tower-http`, where the middleware is imported from, and it seems the issue stemmed from there: I found that there was a recent rewrite of the cors middleware (https://github.com/tower-rs/tower-http/pull/237), I've replaced my usage of `tower_http = { version = "0.2.5", features = ["cors"]}` with `tower_http = { git = "https://github.com/tower-rs/tower-http", features = ["cors"] }`, and it works as expected now:

```
> OPTIONS / HTTP/1.1
> Host: localhost:2790

< HTTP/1.1 200 OK
< access-control-allow-origin: *
< vary: origin
< vary: access-control-request-method
< vary: access-control-request-headers
< access-control-allow-methods: GET,POST
< allow: GET,HEAD
< content-length: 0
< date: Mon, 25 Apr 2022 10:21:27 GMT
```

**Comment by jplatte** [comment]
Yeah, the CORS middleware used to pass through requests without an origin header to the inner service regardless of the request method. There should a new release (0.3.0) of tower-http soon that includes the improvements from the PR you linked.

Next time, please open issues with tower-http on its own repository instead of here though 😉

**Comment by RAnders00** [comment]
I'm sorry about that, it took me a while to figure out the issue was probably with the `tower-http` middleware. Next time I'll know better :)
